fill_segments/1.0/fill_segments.py

```diff
@@ -0,0 +1,348 @@
+#!/usr/bin/python3
+
+"""
+This script will fill emty segments in seg files by creating a dummy segment with log2(CN)=0 and no LOH.
+This is an important step preparing seg files for GISTIC after lifting over from one genome build to another.
+This script assumes that CN is in log2 ratio.
+It requires seg file and chromosome arms file as mandatory inputs. The path to output file also must be specified.
+
+Example:
+
+python3 fill_segments.py --input </path/to/input_file>.seg --chromArm </path/to/chrom_arm_file>.tsv --output </path/to/output_file>.seg
+"""
+
+# import required modules
+import pandas as pd
+import argparse
+import simplejson
+import sys
+
+
+
+def main():
+    # initiate the parser and handle arguments from command line
+    args = parse_args()
+    input_file = args.input
+    output_file = args.output
+    chrom_file = args.chromArm
+
+    # determine the format of input file
+    input_format = input_file[-3:]
+    
+    # check arguments given in command line
+    check_arguments(args, input_format)
+
+    # create a dictionary containing coordinates of chromosome arms
+    arm_chrom = load_chrom_arm(chrom_file)
+    # get the order of chromosomes
+    chrom_order = list(arm_chrom.keys()) + ["buffer"]
+
+    # initialize empty variable for the new segments
+    columns_new = []
+    columns_edges = []
+
+    # initialize list to store all segments, since it is faster than concatenating pd df with large number of segments
+    seg_filled = []
+
+    # assign new vaeriables for ease of referense when creating new segments
+    # '0.00' assigned so that it will be easy to identify fill-in segments, and grep if necessary
+    empty_loh = str("0.0")
+    empty_logr = str("0.00")
+
+
+    # fill segments
+    seg = open(input_file, 'r+') 
+    lines=seg.readlines()
+
+    # first, get header of the file
+    header=lines[0].rstrip("\n").rstrip("\r").split("\t")
+
+    print("Filling missing segments and smoothing centromeres...")
+    # next, go through each segment, skipping the header
+    for i in range(1,len(lines)-1):
+
+        # read 2 segments at a time to compare coordinates of end of previous sefment, and start of the next segments
+        columns_first = (lines[i].rstrip("\n").rstrip("\r")).split("\t")
+        columns_second = (lines[i+1].rstrip("\n").rstrip("\r")).split("\t")
+        if columns_first[5]=='-inf': columns_first[5]=str('-10')
+        if columns_second[5]=='-inf': columns_second[5]=str('-10')
+
+        # insert empty segment from the beginning of chromosome of the first segment in file to complete the telomeric region of first chromosome
+        if i==1: 
+            columns_new = [columns_first[0], columns_first[1], str(arm_chrom[columns_first[1]]['p']['start']), str(int(columns_first[2])-1), empty_loh, empty_logr]
+            seg_filled.append(columns_new)
+            seg_filled.append(columns_first)
+            # deal with fencepost problem
```

Comment:
What is the fencepost problem? 
And if its solved can we add a comment on how, for future .

Reply:
I might refer to it differently. The correct name is probably "off-by-one error". It can be either the first or the last segment since I am reading them in pairs. I basically just included possible scenarios to process the first segment separately from the rest of them.